File: android/src/main/kotlin/io/storecamera/storecamera_photo/storecamera_plugin_media/data/PluginImageBuffer.kt
package io.storecamera.storecamera_photo.storecamera_plugin_media.data

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import androidx.core.graphics.BitmapCompat
import io.storecamera.storecamera_photo.storecamera_plugin_media.BitmapHelper
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.nio.ByteBuffer

class PluginImageBuffer(
        val width: Int?,
        val height: Int?,
        val format: PluginImageFormat,
        private val buffer: ByteArray
) : PluginToMap {

    companion object {
        fun init(format: PluginImageFormat, bitmap: Bitmap?): PluginImageBuffer? =
                bitmap?.let {
                    when(format) {
                        PluginImageFormat.BITMAP -> {
                            val byteCount = BitmapCompat.getAllocationByteCount(it)
                            val buffer: ByteBuffer = ByteBuffer.allocate(byteCount) //Create a new buffer
                            it.copyPixelsToBuffer(buffer) //Move the byte dat
                            PluginImageBuffer(it.width, it.height, format, buffer.array())
                        }
                        PluginImageFormat.JPG -> {
                            var pluginImageBuffer: PluginImageBuffer? = null
                            var out: ByteArrayOutputStream? = null
                            try {
                                out = ByteArrayOutputStream()
                                it.compress(Bitmap.CompressFormat.JPEG, 100, out)
                                pluginImageBuffer = PluginImageBuffer(it.width, it.height, format, out.toByteArray())
                            } catch (e: Exception) {
                                e.printStackTrace()
                            } finally {
                                try {
                                    out?.close()
                                } catch (e: IOException) {
                                    e.printStackTrace()
                                }
                            }
                            return pluginImageBuffer
                        }
                        PluginImageFormat.PNG -> {
                            var pluginImageBuffer: PluginImageBuffer? = null
                            var out: ByteArrayOutputStream? = null
                            try {
                                out = ByteArrayOutputStream()
                                it.compress(Bitmap.CompressFormat.PNG, 100, out)
                                pluginImageBuffer = PluginImageBuffer(it.width, it.height, format, out.toByteArray())
                            } catch (e: Exception) {
                                e.printStackTrace()
                            } finally {
                                try {
                                    out?.close()
                                } catch (e: IOException) {
                                    e.printStackTrace()
                                }
                            }
                            return pluginImageBuffer
                        }
                    }
                }

        fun map(hashMap: HashMap<*, *>?): PluginImageBuffer? {
            val map = hashMap ?: return null
            val format = PluginImageFormat.from(map["format"] as? String) ?: return null
            val buffer = (map["buffer"] as? ByteArray) ?: return null
            val width = map["width"] as? Int
            val height = map["height"] as? Int
            return PluginImageBuffer(width, height, format, buffer)
        }
    }

    fun bitmap(): Bitmap? = when(format) {
        PluginImageFormat.BITMAP -> {
            var result: Bitmap? = null
            this.width?.let {  width ->
                this.height?.let { height ->
                    result = BitmapHelper.createBitmapFromBitmapByteArray(buffer, width, height)
                }
            }
            result
        }
        PluginImageFormat.JPG -> BitmapHelper.decodeBitmapFromByteArray(buffer)
        PluginImageFormat.PNG -> BitmapHelper.decodeBitmapFromPngByteArray(buffer)
    }

    override fun pluginToMap(): Map<String, *> = hashMapOf(
            "format" to format.format,
            "buffer" to buffer).apply {
        width?.let { this["width"] = it }
        height?.let { this["height"] = it }
    }
}

File: ios/Classes/media/data/PluginImageBuffer.swift
//
//  PluginImageBuffer.swift
//  store_camera_plugin_media
//
//  Created by 김경환 on 2020/12/19.
//

class PluginImageBuffer: MediaPluginToMap {

    let width: Int?
    let height: Int?
    let format: PluginImageFormat
    let buffer: Data
    
    init?(_ format: PluginImageFormat, _ image: UIImage?) {
        guard let uiImage = image else {
            return nil
        }
        self.width = Int(uiImage.size.width)
        self.height = Int(uiImage.size.height)
        self.format = format

        switch format {
        case .BITMAP:
            guard let data = uiImage.cgImage?.dataProvider?.data as Data? else {
                return nil
            }
            self.buffer = data
        case .JPG:
            guard let data = uiImage.jpegData(compressionQuality: 1.0) else {
                return nil
            }
            self.buffer = data
        case .PNG:
            guard let data = uiImage.pngData() else {
                return nil
            }
            self.buffer = data
        }
    }
    
    init?(_ dictionary: [String:Any]?) {
        guard let format = PluginImageFormat.init(rawValue: StoreCameraMediaDictionaryHelper.get(dictionary, "format") ?? ""),
              let data: Data = StoreCameraMediaDictionaryHelper.get(dictionary, "buffer") else {
            return nil
        }
        self.format = format
        self.buffer = data
        self.width = StoreCameraMediaDictionaryHelper.get(dictionary, "width")
        self.height = StoreCameraMediaDictionaryHelper.get(dictionary, "height")
    }
    
    func pluginToMap() -> [String : Any] {
        var map: [String:Any] = [
            "format": format.rawValue,
            "buffer": buffer
        ]
        if let w = width {
            map["width"] = w
        }
        if let h = height {
            map["height"] = h
        }
        
        return map
    }
    
    func uiImage() -> UIImage? {
        switch(format) {
        case .BITMAP:
            if let w = width, let h = height {
                return MediaUIImageHelper.createUIImage(buffer, w, h)
            }
            return nil
        case .JPG, .PNG:
            return UIImage(data: buffer)
        }
    }
}
